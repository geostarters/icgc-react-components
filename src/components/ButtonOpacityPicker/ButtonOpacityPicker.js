// @flow

import React from "react";
import PropTypes from "prop-types";
import { AlphaPicker } from "react-color";
import IMAGES from "./images/images";

import styles from "./ButtonOpacityPicker.module.css";

export default class ButtonOpacityPicker extends React.PureComponent {

	state = {

		displayAlphaPicker: false,
		opacity: this.props.opacity

	};

	handleClick = () => this.setState({ displayAlphaPicker: !this.state.displayAlphaPicker });

	handleClose = () => this.setState({ displayAlphaPicker: false });

	handleChange = (opacity) => {

		if (opacity.rgb.a < 0.02) {

			return;

		}
		this.setState(prevState => ({
			...prevState,
			opacity: opacity.rgb.a
		}));
		this.props.handleChangeOpacity(opacity.rgb.a, this.props.index);

	}

	render() {

		const { opacity } = this.state;

		return (
			<div className={ styles.buttonContainer }>
				<div className={ styles.swatch } onClick={ this.handleClick }>
					<div className={ styles.color } style={{backgroundImage: `url(${this.props.image})`}}>
						<div className={ styles.color } style={{background: `rgba(255, 255, 255, ${1 - opacity})`}} />
					</div>
				</div>
				{ this.state.displayAlphaPicker ? <div className={ styles.popover }>
					<div className={ styles.cover } onClick={ this.handleClose }/>
					<div className={ styles.alphaContainer }>
						<AlphaPicker width="300px" color={`rgba(255, 255, 255, ${opacity})`} onChange={ this.handleChange } />
					</div>
				</div> : null }

			</div>

		);

	}

}

ButtonOpacityPicker.propTypes = {
	handleChangeOpacity: PropTypes.func,
	opacity: PropTypes.number,
	index: PropTypes.number,
	image: PropTypes.string
};

ButtonOpacityPicker.defaultProps = {
	handleChangeOpacity: () => {},
	opacity: 1,
	index: 0,
	image: `${IMAGES.opacity}`
};
