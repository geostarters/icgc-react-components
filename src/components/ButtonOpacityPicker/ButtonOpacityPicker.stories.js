//@flow

import React from "react";
import { action } from "@storybook/addon-actions";

import ButtonOpacityPicker from "./ButtonOpacityPicker";

import "semantic-ui-css/semantic.min.css";

export default {
	title: "ButtonOpacityPicker",
	component: ButtonOpacityPicker
};

const Template = args => (<ButtonOpacityPicker {...args}/>);

export const _ButtonOpacityPicker = Template.bind({});
_ButtonOpacityPicker.storyName = "ButtonOpacityPicker";

export const ButtonOpacityPickerOpacity = Template.bind({});
ButtonOpacityPickerOpacity.storyName = "ButtonOpacityPicker opacity";
ButtonOpacityPickerOpacity.args = {
	opacity: 0.5,
	handleChangeOpacity: action("onChange")
};

export const ButtonOpacityPickerImagen = Template.bind({});
ButtonOpacityPickerImagen.storyName = "ButtonOpacityPicker imagen";
ButtonOpacityPickerImagen.args = {
	opacity: 0.5,
	handleChangeOpacity: action("onChange"),
	image: "https://www.instamaps.cat/static/img/map-backgrounds/orto.jpg"
};
