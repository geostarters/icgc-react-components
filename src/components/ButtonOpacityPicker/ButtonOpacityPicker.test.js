// @flow

import React from "react";
import ButtonOpacityPicker from "./ButtonOpacityPicker";
import  { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { AlphaPicker } from "react-color";

describe("ButtonOpacityPicker Component", () => {

	const buttonOpacityPickerProps = {
		opacity: 1,
		index: 0,
		handleChangeOpacity: jest.fn()
	};
	const component = shallow(<ButtonOpacityPicker {...buttonOpacityPickerProps}></ButtonOpacityPicker>);

	it("renders correct properties", () => {

		expect(toJSON(component)).toMatchSnapshot();

	});

	it("has a element with  swatch class", () => {

		expect(component.find(".swatch")).toHaveLength(1);

	});

	it("has the correct opacity in the element", () => {

		expect(component.find(".swatch").find(".color").get(1).props.style).toHaveProperty("background", "rgba(255, 255, 255, 0)");

	});

	it("shows OpacityPicker after click", () => {

		component.find(".swatch").simulate("click");
		expect(component.state("displayAlphaPicker")).toBeTruthy();

	});

	it("call handle function after changing color", () => {

		component.find(AlphaPicker).simulate("change", {rgb: {r: 255, g:255, b: 0, a: 0.5}});
		expect(buttonOpacityPickerProps.handleChangeOpacity).toBeCalled();

	});

});
