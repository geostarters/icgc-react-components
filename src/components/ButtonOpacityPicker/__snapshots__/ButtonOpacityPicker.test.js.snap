// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonOpacityPicker Component renders correct properties 1`] = `
<div
  className="buttonContainer"
>
  <div
    className="swatch"
    onClick={[Function]}
  >
    <div
      className="color"
      style={
        Object {
          "backgroundImage": "url(opacity.jpg)",
        }
      }
    >
      <div
        className="color"
        style={
          Object {
            "background": "rgba(255, 255, 255, 0)",
          }
        }
      />
    </div>
  </div>
</div>
`;
